workflow-runtime/src/main/java/com/squareup/workflow/WorkflowInterceptor.kt

```diff
@@ -0,0 +1,180 @@
+ * different, disjoint times. Each continuous sequence of renderings of a particular workflow type,
+ * with the same key passed to [RenderContext.renderChild], is called an "instance" of that
+ * workflow. The workflow's [StatefulWorkflow.initialState] method will be called at the start of
+ * the instance, and its state will be maintained by the runtime until the instance is finished.
+ * Each instance is identified by the [WorkflowInstance] object passed into the corresponding method
+ * in a [WorkflowInterceptor].
+ *
+ * In addition to the [WorkflowIdentifier] of the type of the workflow being rendered, this object
+ * also knows the [key][WorkflowInstance.renderKey] used to render the workflow and the
+ * [WorkflowInstance] of the [parent][WorkflowInstance.parent] workflow that is rendering it.
+ *
+ * Each instance is also assigned a numerical ID that uniquely identifies the instance over the
+ * life of the entire runtime. This value will remain constant as long as the workflow's parent is
+ * rendering it, and then it will never be used again. If this workflow stops being rendered, and
+ * then starts again, the value will be different.
+ */
+@ExperimentalWorkflowApi
+interface WorkflowInterceptor {
+
+  /**
+   * Called when the instance is starting, before [onInitialState].
+   *
+   * @param workflowScope The [CoroutineScope] that will be used for any side effects the workflow
+   * runs, as well as the parent for any workflows it renders.
+   */
+  // Can perform actions when workflow is cancelled by using Job in the scope,
+  // which will be cancelled when the workflow is.
+  fun onInstanceStarted(
+    workflowScope: CoroutineScope,
+    instance: WorkflowInstance
+  ) = Unit
+
+  /**
+   * Intercepts calls to [StatefulWorkflow.initialState].
+   */
+  fun <P, S> onInitialState(
+    props: P,
+    snapshot: Snapshot?,
+    proceed: (P, Snapshot?) -> S,
+    instance: WorkflowInstance
+  ): S = proceed(props, snapshot)
+
+  /**
+   * Intercepts calls to [StatefulWorkflow.onPropsChanged].
+   */
+  fun <P, S> onPropsChanged(
+    old: P,
+    new: P,
+    state: S,
+    proceed: (P, P, S) -> S,
+    instance: WorkflowInstance
+  ): S = proceed(old, new, state)
+
+  /**
+   * Intercepts calls to [StatefulWorkflow.render].
+   */
+  fun <P, S, O : Any, R> onRender(
+    props: P,
+    state: S,
+    context: RenderContext<S, O>,
+    proceed: (P, S, RenderContext<S, O>) -> R,
+    instance: WorkflowInstance
+  ): R = proceed(props, state, context)
+
+  /**
+   * Intercepts calls to [StatefulWorkflow.snapshotState].
+   */
+  fun <S> onSnapshotState(
+    state: S,
+    proceed: (S) -> Snapshot,
+    instance: WorkflowInstance
+  ): Snapshot = proceed(state)
+
+  /**
+   * Information about the instance of a workflow in the runtime that a [WorkflowInterceptor] method
+   * is intercepting.
+   */
+  @ExperimentalWorkflowApi
+  interface WorkflowInstance {
```

Comment:
Is this name confusing? It's hard to talk about "instances of Workflow classes" vs this. We typically call these "sessions" offline, should we just use that term here? Also cc @square/mobile-foundation-ios for their thoughts on this.